### PR TITLE
Fix leak of reference on WDFREQUEST

### PR DESCRIPTION
**Pull request by Alan-Jowett**
Signed-off-by: Alan Jowett <alan.jowett@microsoft.com>

## Description

If submitting an async request fails, release the ref count acquired before submitting it.

## Testing

CI/CD

## Documentation

No.

Resolves: #1324

### CODE REVIEW

File: ebpfcore/ebpf_drv.c [hunk 1]
```diff
@@ -333,10 +335,11 @@ _ebpf_driver_io_device_control(
 
 Done:
     if (status != STATUS_PENDING) {
-        if (async) {
+        if (wdf_request_ref_acquired) {
             ebpf_assert(status != STATUS_SUCCESS);
             // Async operation failed. Remove cancellable marker.
             (void)WdfRequestUnmarkCancelable(request);
+            WdfObjectDereference(request);
```
Comment (shankarseal) [inline ebpfcore/ebpf_drv.c:342]: Interesting - is this leak only if an async operation completes synchronously?

Comment (Alan-Jowett) [reply ebpfcore/ebpf_drv.c:342]: If it completes synchronously or if it fails.

**Review by dthaler — APPROVED** [review]

**Comment by shankarseal** [comment]
Is a deref call needed in the success case? _ebpf_driver_io_device_control_complete ?

**Comment by Alan-Jowett** [comment]
> Is a deref call needed in the success case? _ebpf_driver_io_device_control_complete ?

Yes. If this is an async call, we do two additional things before calling into the execution context:
1) Mark the request as cancelable.
2) Increment a reference on the request.

The request then ends in one of two possible ways:
1) It completes synchronously (either success or failure)
2) It completes asynchronously (by returning pending).

In both cases, we need to unmark the request as cancelable and release the reference.

**Review by shankarseal — APPROVED** [review]

**Pull request merged**